Fix: Respect user_header_name property for budget selection and user identification

colesmcintosh opened the pull request:
## Title

Fix: Respect user_header_name property for budget selection and user identification

## Relevant issues

https://github.com/BerriAI/litellm/issues/11083

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🐛 Bug Fix
🆕 New Feature

## Changes

### Summary
Extended the `get_end_user_id_from_request_body` function to accept an optional `Request` parameter, enabling the system to respect the `user_header_name` configuration property when extracting end-user IDs from incoming request headers for proper budget tracking and user identification.

### Technical Details

**Modified Files:**
- `litellm/proxy/auth/auth_utils.py`: Extended `get_end_user_id_from_request_body` function signature and logic
- `tests/local_testing/test_auth_utils.py`: Added comprehensive tests for new header-based functionality

**Key Changes:**

1. **Function Signature Extension**: 
   - Extended `get_end_user_id_from_request_body(request_body: dict)` to `get_end_user_id_from_request_body(request_body: dict, request: Optional[Request] = None)`
   - **Fully backwards compatible** - existing code continues to work without changes
   - New header functionality only activated when `request` parameter is provided

2. **Header-Based User ID Extraction**:
   - Added support for `general_settings.user_header_name` configuration
   - When `request` parameter is provided, function checks configured header for user ID
   - Prioritizes custom header value when configured, present, and non-empty
   - Falls back to existing body-based extraction methods
   - Handles edge cases (empty strings, non-string config values, missing headers)

3. **Enhanced Priority Order** (from highest to lowest):
   - **Custom header** (via `user_header_name` config) **[NEW]**
   - Request body `user` field
   - Request body `litellm_metadata.user` field  
   - Request body `metadata.user_id` field

**Test Coverage:**
Added 13 comprehensive test cases covering:
- **Backwards compatibility** - function works with just `request_body` parameter
- Header precedence over body values
- Fallback behavior when header missing/empty/whitespace-only
- Configuration edge cases (None, non-string values)
- Multiple body source fallbacks
- Case-sensitive header matching
- Empty/whitespace header handling


### Benefits
- **Fully backwards compatible** - no breaking changes to existing code
- Enables flexible user identification via custom headers
- Supports proper budget tracking when user IDs are passed via headers
- Robust error handling for misconfigured header names
- Graceful degradation when no request object provided

### Usage Examples
```python
# Backwards compatible usage (existing code continues to work)
user_id = get_end_user_id_from_request_body(request_body)

# New usage with header support
user_id = get_end_user_id_from_request_body(request_body, request)
```

```python
# Example request with custom header
headers = {"X-User-ID": "user-123"}
# Function will now extract "user-123" for budget tracking when request object is provided
```

### Breaking Changes
**None.** This change is fully backwards compatible. Existing call sites continue to work exactly as before without any modifications required.

### Test Results
<img width="919" alt="Screenshot 2025-06-04 at 4 06 26 PM" src="https://github.com/user-attachments/assets/15820142-c9b8-4a42-b9e2-26c3f7870a07" />

All tests pass, including:
- Original functionality tests
- New header-based extraction tests  
- Backwards compatibility verification tests
- Edge case handling tests

colesmcintosh committed:
Refactor get_end_user_id_from_request_body to support user ID retrieval from custom headers and multiple request body formats. Enhance tests to cover various scenarios including header precedence and fallback mechanisms.

colesmcintosh committed:
Merge branch 'BerriAI:main' into user-header-budget-enforcement

colesmcintosh committed:
Refactor get_end_user_id_from_request_body function to accept request_body as the first parameter, improving clarity and flexibility. Update tests for compatibility and add new cases to ensure correct functionality across various request body formats.

colesmcintosh committed:
Update _user_api_key_auth_builder and user_api_key_auth to pass request object to get_end_user_id_from_request_body, enhancing user ID retrieval from request data.

sammcj commented:
If this is the fix for https://github.com/BerriAI/litellm/issues/11083 I'll be wrapped to see this merged!

krrishdholakia commented on litellm/proxy/auth/auth_utils.py:
```diff
@@ -499,17 +499,43 @@ def _has_user_setup_sso():
 
-def get_end_user_id_from_request_body(request_body: dict) -> Optional[str]:
-    # openai - check 'user'
+def get_end_user_id_from_request_body(request_body: dict, request: Optional[Request] = None) -> Optional[str]:
```
can we use `request_headers` here instead of the `request` param? 

this will reduce the scope for a future change accidentally using request.body here (this can slow down calls)

colesmcintosh committed:
Merge branch 'BerriAI:main' into user-header-budget-enforcement

colesmcintosh committed:
refactor(auth_utils.py): update get_end_user_id_from_request_body to accept request_headers instead of request, and adjust related function calls in user_api_key_auth and tests

colesmcintosh committed:
Merge branch 'BerriAI:main' into user-header-budget-enforcement

krrishdholakia commented:
@colesmcintosh looks like this is failing the vertex tests - can you please look into that? 

I confirmed the same test passes on main
<img width="1419" alt="Screenshot 2025-06-05 at 4 48 06 PM" src="https://github.com/user-attachments/assets/80a30162-f2fb-49be-91f3-b60b58febafa" />

colesmcintosh committed:
Merge branch 'BerriAI:main' into user-header-budget-enforcement

colesmcintosh committed:
refactor(tests): update mock request handling in LLM pass-through endpoint tests

- Replaced the Request object with a Mock for better flexibility in testing.
- Enhanced mock setup to include user API key handling and virtual key retrieval.
- Updated test calls to reflect changes in mock request structure and added necessary patches for new dependencies.

colesmcintosh committed:
refactor(vertex_and_google_ai_studio_gemini.py): remove redundant variable declaration for url_context_metadata, linting error

colesmcintosh committed:
Merge branch 'BerriAI:main' into user-header-budget-enforcement
